specs/AddHostObjectToScriptWithOrigins.md

```diff
@@ -0,0 +1,161 @@
+            args.Frame.AddHostObjectToScriptWithOrigins("bridge", new BridgeAddRemoteObject(), origins);
+        }
+        catch (NotSupportedException exception)
+        {
+            MessageBox.Show("Frame.AddHostObjectToScriptWithOrigins failed: " + exception.Message);
+        }
+    };
+}
+```
+
+## Win32 C++
+```cpp
+void SampleClass::SampleMethod()
+{
+    // Register a handler for the FrameCreated event.
+    // This handler can be used to add host objects to the created iframe.
+    CHECK_FAILURE(m_webView->add_FrameCreated(
+        Callback<ICoreWebView2FrameEventHandler>(
+            [this](
+                ICoreWebView2* sender,
+                ICoreWebView2FrameEventArgs* args) -> HRESULT
+    {
+        wil::com_ptr<ICoreWebView2Frame> webviewFrame;
+        CHECK_FAILURE(args->get_Frame(&webviewFrame));
+
+        // Wrap host object
+        VARIANT remoteObjectAsVariant = {};
+        m_hostObject.query_to<IDispatch>(&remoteObjectAsVariant.pdispVal);
+        remoteObjectAsVariant.vt = VT_DISPATCH;
+
+        // Create list of origins which will be checked.
+        // iframe will have access to host object only if its origin belongs
+        // to this list.
+        LONG i = 0;
+        SAFEARRAY* list = SafeArrayCreateVector(VT_BSTR, 0, 1);
+        BSTR origin = SysAllocString(L"https://appassets.example");
+        SafeArrayPutElement(list, &i, origin);
+
+        VARIANT originList = {};
+        originList.parray = list;
+        originList.vt = VT_ARRAY;
+
+        // Add host object to a script for iframe access
+        CHECK_FAILURE(
+            webviewFrame->AddHostObjectToScriptWithOrigins(
+                L"sample", &remoteObjectAsVariant, &originList));
+
+        remoteObjectAsVariant.pdispVal->Release();
+        SysFreeString(origin);
+        SafeArrayDestroy(list);
+
+        return S_OK;
+    }).Get(), &m_frameCreatedToken));
+}
+```
+
+
+# Remarks
+List of origins in AddHostObjectToScript will be treated as following:
+ - empty list - no host object will be added for the iframe;
+ - list with origins - during access to host object from iframe the origin will be
+ checked that it belongs to this list;
+ - list with "*" element - host object will be available for iframe for all origins.
+ We suggest not to use this feature without understanding security implications of
+ giving access to host object from from iframes with unknown origins.
+
+
+# API Notes
+See [API Details](#api-details) section below for API reference.
+
+# API Details
+## IDL
+```c#
+/// WebView2Frame provides direct access to the iframes information and use.
+[uuid(18ae830b-8c83-459f-8a8c-95a1022af774), object, pointer_default(unique)]
+interface ICoreWebView2Frame : IUnknown {
+  /// The name of the iframe from the iframe html tag declaring it.
+  [propget] HRESULT Name([ out, retval ] LPWSTR * name);
+  /// Unique frame identifier.
+  [propget] HRESULT Id([ out, retval ] INT32* frameId);
```

Comment:
We propose to use AddHostObjectToScriptWithOrigins  frame method during frame creation. At this time there is not match information about the frame, so name can be used by developers to distinguish which frame need host objects and which does not. Also I have heard that some applications are using this field. Because name itself is not unique I'm providing Id which is unique. For example: if 5 FrameCreated events are received with the same name - is it the same frame(should not be) or different? Id will provide the answer, because it is unique.

Reply:
Make sense - will keep the name and will remove id.